FormalConjectures/ForMathlib/Combinatorics/AP/Basic.lean

```diff
@@ -0,0 +1,136 @@
+  simp
+  refine Set.ext fun x => ⟨fun h ↦ ?_, fun ⟨n, ⟨_, _⟩⟩ ↦ by interval_cases n <;> simp_all⟩
+  cases h with
+  | inl hl => use 0; simp [zero_nsmul, hl]
+  | inr hr => exact ⟨1, by norm_num, by simp_all [hr, add_assoc]⟩
+
+-- Formalisation note: separate result needed for `ℕ` since this is not covered by
+-- the `AddCommGroup` result above.
+/-- The set `{a, b} : Set ℕ` with `a < b` is an arithmetic progression of length `2` with
+first term `a` and difference `b - a`. -/
+theorem Nat.isAPOfLengthWith_pair {a b : ℕ} (hab : a < b) :
+    Set.IsAPOfLengthWith {a, b} 2 a (b - a) := by
+  let ⟨n, h⟩ := Nat.exists_eq_add_of_lt hab
+  simp [Set.IsAPOfLengthWith, Nat.sub_ne_zero_of_lt hab, h, add_assoc]
+  exact Set.ext fun x => ⟨fun a => by aesop, fun ⟨w, ⟨_, _⟩⟩ => by interval_cases w <;> simp_all⟩
+
+/--
+The predicate that a set `s` is an arithmetic progression of length `l` (possibly infinite).
+This predicate does not assert a specific value for the first term or the difference of the
+arithmetic progression.
+-/
+def Set.IsAPOfLength (s : Set α) (l : ℕ∞) : Prop :=
+  ∃ a d : α, s.IsAPOfLengthWith l a d
+
+/-- Only the empty set is a finite arithmetic progression of length $0$. -/
+theorem Set.IsAPOfLength.zero {α : Type*} [AddCommMonoidWithOne α] [NeZero (1 : α)] {s : Set α} :
+    s.IsAPOfLength 0 ↔ s = ∅ := by
+  simp [Set.IsAPOfLength, Set.IsAPOfLengthWith.zero]
+
+/-- Only singletons are finite arithmetic progressions of length $1$. -/
+theorem Set.IsAPOfLength.one {α : Type*} [AddCommMonoidWithOne α] [NeZero (1 : α)] {s : Set α} :
+    s.IsAPOfLength 1 ↔ ∃ a, s = {a} := by
+  simp [IsAPOfLength, zero_nsmul, IsAPOfLengthWith.one]
+
+theorem Set.isAPOfLength_singleton {α : Type*} [AddCommMonoidWithOne α] [NeZero (1 : α)] (a : α) :
+    Set.IsAPOfLength {a} 1 := by
+  simp [IsAPOfLength.one]
+
+theorem Set.isAPOfLength_pair {α : Type*} [DecidableEq α] [AddCommGroup α] {a b : α} (hab : a ≠ b) :
+    Set.IsAPOfLength {a, b} 2 := by
+  simpa [IsAPOfLength] using ⟨a, b - a, Set.isAPOfLengthWith_pair hab⟩
+
+theorem Nat.isAPOfLength_pair {a b : ℕ} (hab : a < b) :
+    Set.IsAPOfLength {a, b} 2 := by
+  simpa [Set.IsAPOfLength] using ⟨a, b - a, Nat.isAPOfLengthWith_pair hab⟩
+
+/-- We say that a set `s` is free of arithmetic progressions of length `l` if `s` contains no
+non-trivial arithmetic progressions of length `l`. Written as `Set.IsAPOfLengthFree s l`. --/
+def Set.IsAPOfLengthFree (s : Set α) (l : ℕ∞) : Prop :=
+  ∀ t ⊆ s, t.IsAPOfLength l → t.ncard = 1
+
+/-- Every set contains an arithmetic progression of length zero. -/
+@[simp]
+theorem Set.not_isAPOfLengthFree_zero {α : Type*} [AddCommMonoidWithOne α] [NeZero (1 : α)]
+    (s : Set α) : ¬s.IsAPOfLengthFree 0 := by
+  simpa [Set.IsAPOfLengthFree, Set.IsAPOfLength.zero] using fun x ↦ by simp [Ne.symm]
+
+lemma Set.IsAPOfLength.card (s : Set α) (l : ℕ∞) (hs : s.IsAPOfLength l) :
+    ENat.card s = l :=
+  hs.choose_spec.choose_spec.1
```

Comment:
Very big nit but I would put this lemma right after definition (since it's just deconstructing it). And then we should add a `Set.IsAPOfLength.eq` lemma as well for the set equality.